### PR TITLE
Intrepid2: support for (L2) boundary projections

### PR DESCRIPTION
**Pull request by mperego**
This PR adds support for boundary projections and orientations, which is used for prescribing Dirichlet conditions:

- For a 3D cell, we compute the orientations of its faces and edges, not for the cell itself, because they are not needed for prescribing continuity across elements. For a 2D cell, we used to compute the orientations for edges, not for the cell itself.     However, when the 2D cell is the side of a 3D cell, it inherits the orientation of the parent 3D cell,  and it is useful, for side projections, to support the orientation of the 2d cell itself.  Similarly, a 1D side cell inherits the orientation of the 2D parent cell and needs support for its orientation. Here we provide such support. We also add orientations for 1D and 2D HVol basis, which are used, e.g., when HDiv functions are restricted to the boundary (in the normal direction).
- We add functions for mapping data from the physical space to the reference space, both form physical cells to reference cells and from physical sides to reference sides. L2 projection is modify to support side orientations.
- added support for boundary info in CartesianConnManager and fixed bug in edges numbering.
- modified `test_fe_projection` in Panzer Dof Manager to test boundary projections as well.

We also added additional modifications including
- fixed use of HVol basis: Until now it was possible to create an HVol basis using HGRAD basis classes with `Gaussian` `PointType`. That capability has been removed because it was confusing and it wrongly assigned the space `FUNCTION_SPACE_HGRAD` to HVol basis. Now only HVol basis classes are used for creating HVol basis.
- made all Projection tests (in particular the DeRham ones) UVM free
- fixed a bug in HCurl L2 projections on tets

@trilinos/intrepid2 
@trilinos/panzer 

## Testing
The functions added are tested by modified Intrepid2 projections DeRham tests and a Panzer Dof Manager projection test.

### CODE REVIEW

**Review by rppawlo — APPROVED** [review]
@mperego - looks good to me! Did you get a chance to test these changes with a UVM=OFF build? PR testing currently doesn't protect UVM=OFF builds but we will need this capability working for an upcoming milestone.

File: packages/intrepid2/src/Cell/Intrepid2_CellToolsDefNodeInfo.hpp [hunk 1]
```diff
@@ -451,6 +451,96 @@ namespace Intrepid2 {
...
+    // Storage for constant reference edge tangent: rank-1 (D) arrays
+    const ordinal_type dim = parentCell.getDimension();
+    auto vcprop = Kokkos::common_view_alloc_prop(edgeTangents);
+    using common_value_type = typename decltype(vcprop)::value_type;
+    Kokkos::DynRankView< common_value_type, DeviceType > refEdgeTan ( Kokkos::view_alloc("CellTools::getPhysicalEdgeTangents::refEdgeTan", vcprop), edgeTangents.extent(0), dim);
+
+    const auto edgeMap = RefSubcellParametrization<DeviceType>::get(1, parentCell.getKey());
+
+    using FunctorType = FunctorCellTools::F_refEdgeTangent<decltype(refEdgeTan),decltype(worksetEdgeOrds),decltype(edgeMap)>;
+    Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpaceType>(0,refEdgeTan.extent(0)), FunctorType(refEdgeTan, worksetEdgeOrds, edgeMap) );
```
[the remark this answers is not in the document]
Comment (mperego) [reply packages/intrepid2/src/Cell/Intrepid2_CellToolsDefNodeInfo.hpp:539]: I think you are right. Thanks for catching this. I added the fence.

File: packages/intrepid2/src/Cell/Intrepid2_CellToolsDefNodeInfo.hpp [hunk 2]
```diff
@@ -517,16 +607,122 @@ namespace Intrepid2 {
...
+
+    auto vcprop = Kokkos::common_view_alloc_prop(faceTanU);
+    using common_value_type = typename decltype(vcprop)::value_type;
+    Kokkos::DynRankView< common_value_type, DeviceType > refFaceTanU ( Kokkos::view_alloc("CellTools::getPhysicalFaceTangents::refFaceTanU", vcprop), faceTanU.extent(0), dim);
+    Kokkos::DynRankView< common_value_type, DeviceType > refFaceTanV ( Kokkos::view_alloc("CellTools::getPhysicalFaceTangents::refFaceTanV", vcprop), faceTanV.extent(0), dim);
+
+    const auto faceMap = RefSubcellParametrization<DeviceType>::get(2, parentCell.getKey());
+
+    using FunctorType = FunctorCellTools::F_refFaceTangents<decltype(refFaceTanU),decltype(worksetFaceOrds),decltype(faceMap)>;
+    Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpaceType>(0,refFaceTanU.extent(0)), FunctorType(refFaceTanU, refFaceTanV, worksetFaceOrds, faceMap) );
```
[the remark this answers is not in the document]
Comment (mperego) [reply packages/intrepid2/src/Cell/Intrepid2_CellToolsDefNodeInfo.hpp:709]: added the fence.

**Comment by kyungjoo-kim** [comment]
@mperego @rppawlo it does not need uvm setting for testing intrepid2. our unit tests always test the code without uvm.

https://github.com/trilinos/Trilinos/blob/0bd35b1d0298b79ab6e18619e61fa364c84a6267/packages/intrepid2/unit-test/Projection/CMakeLists.txt#L88

**Comment by mperego** [comment]
> @mperego @rppawlo it does not need uvm setting for testing intrepid2. our unit tests always test the code without uvm.
> 
> https://github.com/trilinos/Trilinos/blob/0bd35b1d0298b79ab6e18619e61fa364c84a6267/packages/intrepid2/unit-test/Projection/CMakeLists.txt#L88

Right. But I also modified Panzer Dof-Mgr anf in that case you need to build with UVM off.

File: packages/intrepid2/src/Projection/Intrepid2_ProjectionToolsDefL2.hpp [hunk 3]
```diff
@@ -407,7 +452,23 @@ ProjectionTools<DeviceType>::getL2EvaluationPoints(typename BasisType::ScalarVie
   if(numVols > 0) {
     auto pointsRange = ePointsRange(dim, 0);
     auto cellEPoints = Kokkos::create_mirror_view_and_copy(MemSpaceType(),projStruct->getEvalPoints(dim,0,ePointType));
-    RealSpaceTools<DeviceType>::clone(Kokkos::subview(ePoints, Kokkos::ALL(), pointsRange, Kokkos::ALL()), cellEPoints);
+    if(dim == 3)
+      RealSpaceTools<DeviceType>::clone(Kokkos::subview(ePoints, Kokkos::ALL(), pointsRange, Kokkos::ALL()), cellEPoints);
+    else { //if the cell is a side also internal points need orientation
+      const auto topoKey = refTopologyKey(dim,0);
+      RealSpaceTools<DeviceType>::clone(Kokkos::subview(ePoints, Kokkos::ALL(), pointsRange, Kokkos::ALL()), cellEPoints);
+        Kokkos::parallel_for
```
Comment (CamelliaDPG) [inline packages/intrepid2/src/Projection/Intrepid2_ProjectionToolsDefL2.hpp:460]: Maybe want a fence here?

**Review by CamelliaDPG — DISMISSED** [review]
@mperego, looks good to me.  I noted a few additional places where you might want/need to add fences, as well as a few places where the Doxygen didn't seem to match up with the code it was documenting.  Please fix these things if you agree; I defer to you if you disagree (in particular, some fence suggestions might not be needed).  Thanks!

File: packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp [hunk 4]
```diff
@@ -826,6 +952,65 @@ namespace Intrepid2 {
...
+
+        It computes
+        \f[
+        \widehat{\bf u}_p = \sqrt{\text{det}(T^T T)}\; ({\bf u}_p \cdot {\bf n}_p )
+        \f]
+        here \f$ T \f$ is a 3x2 matrix, whose columns are the two physical side tangents, computed differentiating the parameterization of the physical side
+        (see  <var>CellTools::getPhysicalFaceTangents</var> and var>CellTools::getPhysicalEdgeTangents</var>). Note \text{det}(T^T T)} is equivalent to the square
+        of the norm of the physical normal computed with <var>CellTools::getPhysicalSideNormal</var>.
+
+        \param  outputVals      [out] - HVOL data in reference space, indexed by (C,P).
```
[the remark this answers is not in the document]
Comment (mperego) [reply packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp:999]: No, the output, normal direction of HDIV at the boundary, is an HVOL data on the boundary manifold. Does it make sense?

File: packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp [hunk 5]
```diff
@@ -826,6 +952,65 @@ namespace Intrepid2 {
...
+        \param  metricTensorDet [in]  - determinant of the metric tensor \f$T^T T\f$, indexed by (C,P).
+        \param  inputVals       [in]  - cross product of a HCURL function with the unit outer normals to the sides, indexed by (C,P,3)
```
[the remark this answers is not in the document]
Comment (mperego) [reply packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp:1001]: You are right. I'll fix this.

File: packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp [hunk 6]
```diff
@@ -826,6 +952,65 @@ namespace Intrepid2 {
...
+    mapHDivDataFromPhysToRef(       Kokkos::DynRankView<outputValValueType,       outputValProperties...>       outputVals,
+                              const Kokkos::DynRankView<jacobianInverseValueType, jacobianInverseProperties...> jacobianInv,
+                              const Kokkos::DynRankView<jacobianDetValueType,     jacobianDetProperties...>     jacobianDet,
+                              const Kokkos::DynRankView<inputValValueType,        inputValProperties...>        inputVals );
+
+
+    /** \brief Transformation of HDIV data from physical side to reference side.
+        It takes the input vector \f$ ({\bf u}_p \cdot {\bf n}_p) \f$ defined on physical sides and maps it into \f$ \widehat{\bf u}_p \f$ on the reference side.
+        \f$ {\bf u}_p \f$ is a HDIV function evaluated at points \f$ {\bf x}_p \f$, \f$ {\bf n}_p \f$ the outer unit normal to the sides at point \f${\bf x}_p \f$ and  \f$ \widehat{\bf u}_p \f$ is a
+        HVOL data evaluated at reference points \f$ \widehat{\bf x}_p \f$.
```
[the remark this answers is not in the document]
Comment (mperego) [reply packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp:989]: I think this is correct. \f$ \widehat{\bf u}_p \f$ is an HVOL function.

File: packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp [hunk 7]
```diff
@@ -923,7 +1108,31 @@ namespace Intrepid2 {
...
+        defined on the reference cell and indexed by (C,P).
+
+        Computes the map
+        \f$\widehat{u} = J\, u\f$
+
+        where \f$u\f$ is the input HVOL data in the physical space, evaluated at points \f$ x_p \f$, \f$\widehat{u}\f$ is the output HVOL data on the reference cell
+        evaluated at the reference points \f$ \widehat{x}_p\f$ and \f$ x_p\f$ is the point on physical side corresponding to the the point \f$ \widehat{x}_p \f$ on the reference cell, through the map \f$F\f$.
+        \f$J\f$ is the determinant of the Jacobian of the map \f$F\f$.
+
+        \param  outputVals   [out] - Output array with reference HDIV data, indexed by (C,P)
```
[the remark this answers is not in the document]
Comment (mperego) [reply packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp:1123]: right, I'll fix this.

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefCoeffMatrix_HVOL.hpp [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Kyungjoo Kim  (kyukim@sandia.gov), or
+//                    Mauro Perego  (mperego@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+
+/** \file   Intrepid2_OrientationToolsDefCoeffMatrix_HGRAD.hpp
```
[the remark this answers is not in the document]
Comment (mperego) [reply packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefCoeffMatrix_HVOL.hpp:44]: thanks, I'll fix this.

**Comment by mperego** [comment]
@CamelliaDPG @rppawlo I'm going to add the fences you suggested, but I'm not sure why we need to do so. We are using the same execution space and we are not going back an forth from device and host, if I understand correctly. If these fences are required, then similar fences would be required in other functions in the `CellTools` and `FunctionSpaceTools` functions.

**Comment by mperego** [comment]
@rppawlo @CamelliaDPG do you understand what's going on with the intel build?

**Comment by e10harvey** [comment]
> No idea. @trilinos/framework - can you help us understand what is happening in the intel build? The configure was successful, but no results are posted for the build and install.

The build timed out. Framework has a ticket open to submit these partial results to CDash upon time out via a try-catch-finally block. I will try to get this prioritized.

CC: @ndellingwood, @jwillenbring

**Comment by CamelliaDPG** [comment]
> @CamelliaDPG @rppawlo I'm going to add the fences you suggested, but I'm not sure why we need to do so. We are using the same execution space and we are not going back an forth from device and host, if I understand correctly. If these fences are required, then similar fences would be required in other functions in the `CellTools` and `FunctionSpaceTools` functions.

@mperego, I believe we need fences anytime we require a potentially parallel execution block to have completed its work prior to the execution of whatever follows the fence; if you have two `parallel_for`s executing on device for example, and the second one depends on the completion of the first, then you need a fence.

It is entirely possible that we've missed fences in other Intrepid2 functions.  It's something we could easily get away with without knowing it for quite a while.  We may want to adopt a policy that all user-facing calls include a fence on the relevant execution space at the end; I don't think we follow that now, and it would make it a lot easier both for us and for our users to get this sort of thing right.  (It does incur a performance cost, but I would think it would be negligible in almost every case.)

File: packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/src/Discretization/FunctionSpaceTools/Intrepid2_FunctionSpaceTools.hpp:999]: That makes sense.  Thanks!

**Review by CamelliaDPG — APPROVED** [review]

**Comment by rppawlo** [comment]
If the two calls use the same cuda stream, then you don't need to fence as the kernels are guaranteed to execute one after the other. The main reason we have so many fences is for UVM use. You can't access UVM from the host and device at the same time, so the fence was a safety check. I was not sure if fences were needed there, I know we had to fence similar calls in empire, but we use a lot more uvm in that case.

**Comment by mperego** [comment]
@CamelliaDPG  @rppawlo  @kyungjoo-kim thanks for your explanations. I merged this branch so that we can move on, we'll reassess the need of fences later.

**Pull request merged**